src/amo/components/Addon/index.js

```diff
@@ -73,9 +80,21 @@ export class AddonBase extends React.Component {
   }
 
   componentWillMount() {
-    const { addon, dispatch } = this.props;
+    const { addon, dispatch, errorHandler, params } = this.props;
+
+    if (addon) {
+      dispatch(setViewContext(addon.type));
+    } else {
+      dispatch(fetchAddon({ slug: params.slug, errorHandler }));
+    }
+  }
 
-    dispatch(setViewContext(addon.type));
+  componentWillReceiveProps({ addon: newAddon }) {
+    const { addon: oldAddon, dispatch } = this.props;
```

Comment:
I don't think we should add workarounds for standard JavaScript syntax. It is what it is and [it's well documented](). It will be around for a long time :) We can help teach contributors if they get stuck with it.